### PR TITLE
Basic abstraction layer implementation

### PR DESCRIPTION
**Pull request by OlegOLK**
Introduce core project with abstractions inside.
Add mongo DB-related project with repository implementations and type mappings to DTO and back.
Add external clients project that will communicate with external data to fulfill our DB dataset

### CODE REVIEW

File: src/PortfolioTrackerVNext/PortfolioTrackerVNext.Core/BalanceSnapshot.cs [hunk 1]
```diff
@@ -0,0 +1,41 @@
+﻿using Ardalis.GuardClauses;
+using System;
+
+namespace PortfolioTrackerVNext.Core
+{
+    /// <summary>
+    /// Defines base class for snapshot of the balance for specific date.
+    /// Posible placese where it can be handy: account, wallet, portfolio level snapshots
+    /// </summary>
+    public sealed class BalanceSnapshot
```
[the remark this answers is not in the document]
Comment (quadhd) [reply src/PortfolioTrackerVNext/PortfolioTrackerVNext.Core/BalanceSnapshot.cs:10]: 2nd thought. depositedAmount and withdrawnAmount are never used separately. They can be combined into one value relocateAmount = depositedAmount - withdrawnAmount

File: src/PortfolioTrackerVNext/PortfolioTrackerVNext.Database.Mongo/Repositories/Snapshots/ResilentSnapshotRepository.cs [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+        /// <summary>
+        /// Executes method with all configured policies on board, e.g.: retry, circuit-breaker etc.
+        /// </summary>
+        /// <param name="accountId"></param>
+        /// <returns></returns>
+        public async Task<IEnumerable<BalanceSnapshot>> GetAllSnapshotsForAccountAsync(string accountId)
+        {
+            return await _policy.ExecuteAsync(async () =>
+                await _spanshotRepository.GetAllSnapshotsForAccountAsync(accountId).ConfigureAwait(false))
```
Comment (quadhd) [inline src/PortfolioTrackerVNext/PortfolioTrackerVNext.Database.Mongo/Repositories/Snapshots/ResilentSnapshotRepository.cs:30]: As we now have calculation of snapshots for single wallet let's call it `GetAllSnapshotsForWalletAsync`. Combining Wallet snapshots data into Account snapshots is something that we need to implement later.

File: src/PortfolioTrackerVNext/PortfolioTrackerVNext.Database.Mongo/Models/V1/Snapshots/Snapshot.cs [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+using PortfolioTrackerVNext.Database.Mongo.Models.Interfaces;
+using PortfolioTrackerVNext.Database.Mongo.Models.Interfaces.Relations;
+using System;
+
+namespace PortfolioTrackerVNext.Database.Mongo.Models.Snapshots
+{
+    public class Snapshot : IDbEntity, IAccountRelatedEntity
+    {
+        public ObjectId Id { get; } = ObjectId.GenerateNewId();
+        public ObjectId AccountId { get; set; }
```
Comment (quadhd) [inline src/PortfolioTrackerVNext/PortfolioTrackerVNext.Database.Mongo/Models/V1/Snapshots/Snapshot.cs:13]: And we need `walletId` here.

File: src/PortfolioTrackerVNext/PortfolioTrackerVNext.Core/Transaction.cs [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+        public TransactionType Type { get; }
+        public string SellAsset { get; }
+        public decimal SellAmount { get; }
+        public string BuyAsset { get; }
+        public decimal BuyAmount { get; }
+        public decimal Fee { get; }
+        public string FeeAsset { get; }
+        public TransactionSource Source { get; }
+        public decimal? CostInReferenceAsset { get; }  // USDT
+
```
Comment (quadhd) [inline src/PortfolioTrackerVNext/PortfolioTrackerVNext.Core/Transaction.cs:32]: let's add `CostInReferenceAsset2`

File: src/PortfolioTrackerVNext/PortfolioTrackerVNext.Core/Transaction.cs [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+        //We can acutally move this to the separate collection
+        //with some kind of relatation
+        //For groupping multiple Transaction inside one circle (buy+sell)
+        public decimal? PnlInReferenceAsset { get; }  // 100%   1-1 
+        public bool IsBuyer { get; }
```
Comment (quadhd) [inline src/PortfolioTrackerVNext/PortfolioTrackerVNext.Core/Transaction.cs:37]: I don't think we need `IsBuyer`

File: src/PortfolioTrackerVNext/PortfolioTrackerVNext.Database.Mongo/Models/V1/Account/AssetBalance.cs [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+        /// <summary>
+        /// Reference to <see cref="Snapshot"/>
+        /// </summary>
+        public ObjectId AccountSnapshotId { get; set; }
+
+        /// <summary>
+        /// Asset symbol, e.g.: BTC, USDT, XPX, ETH
+        /// This value should be normalized.
+        /// </summary>
+        public string Symbol { get; set; }
```
Comment (quadhd) [inline src/PortfolioTrackerVNext/PortfolioTrackerVNext.Database.Mongo/Models/V1/Account/AssetBalance.cs:21]: Let's call it AssetName

File: src/PortfolioTrackerVNext/PortfolioTrackerVNext.Core/Transaction.cs [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+        public string SellAsset { get; }
+        public decimal SellAmount { get; }
+        public string BuyAsset { get; }
+        public decimal BuyAmount { get; }
+        public decimal Fee { get; }
+        public string FeeAsset { get; }
+        public TransactionSource Source { get; }
+        public decimal? CostInReferenceAsset { get; }  // USDT
+
+        //We can acutally move this to the separate collection
```
Comment (quadhd) [inline src/PortfolioTrackerVNext/PortfolioTrackerVNext.Core/Transaction.cs:33]: Yes, But that would be collection with 2 fields( we also would need PnlInReferenceAsset2). Not sure if it worth it. But this fields would only be populated for sells

File: src/PortfolioTrackerVNext/PortfolioTrackerVNext.Database.Mongo/Models/V1/Trades/MongoTransaction.cs [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+        public string SellSymbol { get; set; }
+        public decimal SellAmount { get; set; }
+        public string BuySymbol { get; set; }
+        public decimal BuyAmount { get; set; }
+        public decimal Fee { get; set; }
+        public string FeeSymbol { get; set; }
+        public TransactionSource Source { get; set; }
+        public bool IsBuyer { get; set; }
+        public decimal? CostInReferenceSymbol { get; set; }
+        public decimal? PnlInReferenceSymbol { get; set; }
```
Comment (quadhd) [inline src/PortfolioTrackerVNext/PortfolioTrackerVNext.Database.Mongo/Models/V1/Trades/MongoTransaction.cs:35]: Change Symbol to Asset

File: src/PortfolioTrackerVNext/PortfolioTrackerVNext.Database.Mongo/Models/V1/Snapshots/Snapshot.cs (hunk 3 above)
Comment (quadhd) [reply src/PortfolioTrackerVNext/PortfolioTrackerVNext.Database.Mongo/Models/V1/Snapshots/Snapshot.cs:13]: Ignore. I see now that `Reference` can point to WalletId

File: src/PortfolioTrackerVNext/PortfolioTrackerVNext.Database.Mongo/Repositories/Snapshots/ResilentSnapshotRepository.cs (hunk 2 above)
Comment (OlegOLK) [reply src/PortfolioTrackerVNext/PortfolioTrackerVNext.Database.Mongo/Repositories/Snapshots/ResilentSnapshotRepository.cs:30]: This is an example of snapshot API, we will extend it in the future

File: src/PortfolioTrackerVNext/PortfolioTrackerVNext.ExternalClients/Binance/Repositories/BinanceTransactionRepository.cs [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+            // Write new transactions to DB
+            var newBinanceTransactions = list.Select(x => x.ConvertBinanceTradeToTransaction(_binanceStaticData));
+
+            if (!newBinanceTransactions.Any())
+            {
+                return transactions;
+            }
+            newBinanceTransactions = await _transactionRepository.CreateWalletTransactionsAsync(accountId, walletId, newBinanceTransactions).ConfigureAwait(false);
+
+            var res = Sorter.MergeSortedEnumerable(newBinanceTransactions, aggregated, transaction => transaction.DateTime);
```
[the remark this answers is not in the document]
Comment (OlegOLK) [reply src/PortfolioTrackerVNext/PortfolioTrackerVNext.ExternalClients/Binance/Repositories/BinanceTransactionRepository.cs:75]: I think about removing that method at all, all we need is just sort by DateTime. There is no need in that custom functiion

File: src/PortfolioTrackerVNext/PortfolioTrackerVNext.Core/Transaction.cs (hunk 5 above)
Comment (OlegOLK) [reply src/PortfolioTrackerVNext/PortfolioTrackerVNext.Core/Transaction.cs:37]: This is not only about BTCUSDT. This thing needed to identify trading pair to group transactions by  that symbol and then pick up latest trade ID for pair, e.g.:
BTCUSDT  - 10321
ETHUSDT - 31
XPXBTC - 32132

File: src/PortfolioTrackerVNext/PortfolioTrackerVNext.ExternalClients/Binance/Repositories/BinanceTransactionRepository.cs (hunk 9 above)
[the remark this answers is not in the document]
Comment (quadhd) [reply src/PortfolioTrackerVNext/PortfolioTrackerVNext.ExternalClients/Binance/Repositories/BinanceTransactionRepository.cs:75]: Ok

File: src/PortfolioTrackerVNext/PortfolioTrackerVNext.Database.Mongo/Repositories/Snapshots/ResilentSnapshotRepository.cs (hunk 2 above)
Comment (quadhd) [reply src/PortfolioTrackerVNext/PortfolioTrackerVNext.Database.Mongo/Repositories/Snapshots/ResilentSnapshotRepository.cs:30]: Agree

**Pull request merged**